diff --git a/node_modules/oae-config/lib/api.js b/node_modules/oae-config/lib/api.js
index ded34e6..37aaca0 100644
--- a/node_modules/oae-config/lib/api.js
+++ b/node_modules/oae-config/lib/api.js
@@ -147,6 +147,8 @@ var filterModulesByConfig = function(modules, callback) {
 
     var done = 0;
 
+    console.log('filterModulesByConfig');
+
     /**
      * Filters out modules that don't have a configuration file in the config directory
      * @param {Number} i index of the module to filter next
@@ -169,6 +171,7 @@ var filterModulesByConfig = function(modules, callback) {
     };
 
     for (var i = 0; i < modules.length; i++) {
+        console.log('Filtering %s', modules[i]);
         doFilter(modules[i]);
     }
 };
@@ -181,9 +184,12 @@ var filterModulesByConfig = function(modules, callback) {
 var getModules = module.exports.getModules = function(callback) {
     callback = callback || function(){};
 
+    console.log('getModules');
+
     // Read directories starting with oae- and get their configuration files.
     // If there's no configuration file the directory is skipped.
     OAE.getAvailableModules(function(finalModules) {
+        console.log('getAvailableModules.callback');
         configurableModules = {};
         filterModulesByConfig(finalModules, callback);
     });
